synapse/handlers/event_auth.py

```diff
@@ -88,8 +88,69 @@ def compute_auth_events(
             if auth_ev_id:
                 auth_ids.append(auth_ev_id)
 
+        # If the current room is using restricted join rules, an additional event
+        # must be included to assert that the server has the right to authorise
+        # a join event.
+        if event.type == EventTypes.Member:
+            if await self.has_restricted_join_rules(
+                current_state_ids, event.room_version
+            ):
```

Comment:
We'll also need to update `auth_types_for_event` somehow. I'm not really sure how best to do that though.

Reply:
I wonder if we should add a field to the join event of `joined_via_user: <user_id>`, which corresponds to the user ID in the auth events. That way the set of auth types for a given event is still well specified.

Reply:
Let's leave this conversation to the MSC?